### DISCUSSION TITLE
Can I leverage the monitoring stage to use it while running tests on an image?

### DISCUSSION THREAD

**Discussion by bonymoloparambil** [slimtoolkit/slim#709] [General]
Can I use the slim such a way that the _monitoring_ stage can be performed while I run functional tests (With test tool) on an image so that it will cover all the aspects before deciding on what packages to be removed?

**Comment by kcq** [comment]
@bonymoloparambil yes, it's possible. The exact flag combination depends on what kind of tests you have and how you need/want to run them.

First, you'll need to adjust how the tool knows that you are done running your tests and it can continue its data processing. For that you need to set the `--continue-after` flag to `signal` and when you are done running your tests you / your script needs to send the `USR1` signal to the the `slim` process. There are other flag values you might try especially if you are going through the process manually to make sure it all works as expected. For example, you can set the value to `enter` and then instead of sending the `USR1` signal to the tool you press the enter key on your keyboard when you are running the tool.

If you have a web app or a microservice and you want to run your network-based tests then you need to configure the network/http port used by the tool in your host environment. There are two flags for it. The `--publish-port` flag is similar to the `-p` flag in Docker and its value format options are similar too. With that port configured your network-based tests can point to that port number. The other flag is `--publish-exposed-ports` and it's used to map the ports declared in your Dockerfile with the EXPOSE instruction to the same port numbers in your host environment. Let's say your Dockerfile has `EXPOSE 8080` then using `--publish-exposed-ports` will  tell the tool to use 8080 in the host environment for your container, so your network-based test can point to port 8080.

If you have tests that need to run inside of your container then there are a couple of options. For very basic tests where you just have a simple shell script with a bunch of commands to execute then you can use the `--exec-file` flag. It's similar to using `docker exec` and running those commands specified in the file you'd pass to `--exec-file`. You can also use `docker exec` directly to run any commands you want in the temporary container the tool creates (you'll see the container name in console), but you'll need to do a bit of extra work to get that container name, which you can script (the temporary containers the tool creates all have the same prefix).

Another option is to mount your tests into the temporary container the tool creates and set the entrypoint for the container to the test execution script file name. For that you'd use the `--mount` and `--entrypoint` flags (there's also `--cmd`). For example, `slim build --show-clogs=true --entrypoint /tests/runtests.sh --mount $(pwd)/tests/:/tests/ YOUR_IMAGE`
